ament_mypy/ament_mypy/main.py

```diff
@@ -0,0 +1,246 @@
+                       filenames: List[str],
+                       elapsed: float) -> str:
+    xml = textwrap.dedent("""\
+        <?xml version="1.0" encoding="UTF-8"?>
+        <testsuite
+        name="{test_name:s}"
+        tests="{test_count:d}"
+        failures="{error_count:d}"
+        time="{time:s}"
+        >
+    """).format(
+                test_name=testname,
+                test_count=max(len(errors), 1),
+                error_count=len(errors),
+                time='{:.3f}'.format(round(elapsed, 3))
+    )
+
+    if errors:
+        # report each mypy error/warning as a failing testcase
+        for error in errors:
+            pos = ''
+            if error.group('lineno'):
+                pos += ':' + str(error.group('lineno'))
+                if error.group('colno'):
+                    pos += ':' + str(error.group('colno'))
+            xml += _dedent_to("""\
+                <testcase
+                    name={quoted_name}
+                    classname="{test_name}"
+                >
+                    <failure message={quoted_message}/>
+                </testcase>
+                """, '  ').format(
+                    quoted_name=quoteattr(
+                        '{0[type]} ({0[filename]}'.format(error) + pos + ')'),
+                    test_name=testname,
+                    quoted_message=quoteattr('{0[msg]}'.format(error) + pos)
+                )
+    else:
+        # if there are no mypy problems report a single successful test
+        xml += _dedent_to("""\
+            <testcase
+              name="mypy"
+              classname="{}"
+              status="No problems found"/>
+            """, '  ').format(testname)
+
+    # output list of checked files
+    xml += "  <system-out>Checked files:{escaped_files}\n  </system-out>\n".format(
+        escaped_files=escape(''.join(['\n* %s' % f for f in filenames]))
+    )
+
+    xml += '</testsuite>\n'
+    return xml
+
+
+def _get_files(paths: List[str]) -> List[str]:
+    files = []
+    for path in paths:
+        if os.path.isdir(path):
+            for dirpath, dirnames, filenames in os.walk(path):
+                if 'AMENT_IGNORE' in filenames:
+                    dirnames[:] = []
+                    continue
+                # ignore folder starting with . or _
+                dirnames[:] = [d for d in dirnames if d[0] not in ['.', '_']]
+                dirnames.sort()
+
+                # select files by extension
+                for filename in sorted(filenames):
+                    if filename.endswith('.py'):
+                        files.append(os.path.join(dirpath, filename))
+        elif os.path.isfile(path):
+            files.append(path)
+    return [os.path.normpath(f) for f in files]
+
+
+def _get_errors(report_string: str) -> List[Match]:
+    return list(re.finditer(r"^(?P<filename>[^:]+):((?P<lineno>\d+):)?((?P<colno>\d+):)?\ (?P<type>error|warning):\ (?P<msg>.*)$", report_string, re.MULTILINE))  # noqa
```

Comment:
Since `mypy` is invoked with `--show-error-context` it would be good to also extract the helpful information of the previous `note:` line.

Reply:
Should that be added to the xml output as another <testcase> element, or part of the failure in the associated testcase? The schema has a few places where it could go. https://www.ibm.com/support/knowledgecenter/en/SSUFAU_1.0.0/com.ibm.rsar.analysis.codereview.cobol.doc/topics/cac_useresults_junit.html

Reply:
I'd think putting it in the text for the failure itself would be ideal.

Reply:
Handled in 27367a since notes can also appear afterward with other info (case: linking to typesheds"). They are now in their own testcase entry.